### ISSUE TITLE
client.fork.fork() test fails in some envorinments

### ISSUE THREAD

**Issue by p-alik** [gearman/gearmand#44]
@esabol mentioned the issue in the [group](https://groups.google.com/d/msg/gearman/fYgPInGHgXQ/zeRi4EdbBgAJ).

I tested it on Ubuntu
- 14.04 gcc 4.8.4 test passed
- 16.04 5.4.0 test failed

Putting blamed [ASSERT_EQ](https://github.com/gearman/gearmand/blob/master/tests/libgearman-1.0/fork.cc#L109) in try block 
```diff
diff --git a/tests/libgearman-1.0/fork.cc b/tests/libgearman-1.0/fork.cc
index 87d5a2c..2d5a88a 100644
--- a/tests/libgearman-1.0/fork.cc
+++ b/tests/libgearman-1.0/fork.cc
@@ -105,8 +105,12 @@ test_return_t check_client_fork_TEST(void*)
             }
             ASSERT_EQ(EXIT_SUCCESS, exit_code);
           }
-
-          ASSERT_EQ(GEARMAN_SUCCESS, gearman_client_echo(&client, test_literal_param("parent forked")));
+          try{
+                ASSERT_EQ(GEARMAN_SUCCESS, gearman_client_echo(&client, test_literal_param("parent forked")));
+          } catch (const libtest::__test_result& err)
+          {
+                  Error << gearman_client_error(&client);
+          }
         }
     }
     return TEST_SUCCESS;
```
shows the reason:
`tests/libgearman-1.0/fork.cc:113: in check_client_fork_TEST() pid(31365) flush(Broken pipe) lost connection to server during send -> libgearman/connection.cc:903: pid(31365)`

Paradoxically neither `gearman_client_echo` in try block nor  `ASSERT_EQ` failed if I change `fork.cc` in this way:
```diff
index 87d5a2c..0cb174f 100644
--- tests/libgearman-1.0/fork.cc
+++ tests/libgearman-1.0/fork.cc
@@ -105,7 +105,12 @@ test_return_t check_client_fork_TEST(void*)
             }
             ASSERT_EQ(EXIT_SUCCESS, exit_code);
           }
-
+          try{
+                gearman_client_echo(&client, test_literal_param("parent forked"));
+          } catch (const libtest::__test_result& err)
+          {
+                  Error << gearman_client_error(&client);
+          }
           ASSERT_EQ(GEARMAN_SUCCESS, gearman_client_echo(&client, test_literal_param("parent forked")));
         }
     }
```

**Comment by esabol** [comment]
The test fails on RHEL/CentOS 7.2 when compiled with gcc 4.8.5.

**Comment by SpamapS** [comment]
This is a racey test. The result should be in the test log. You can also just run the test directly. It's not platform specific, it's just a badly written test.

**Comment by SpamapS** [comment]
Been doing some analysis. So what happens is this:

1) child forks, which ends up with the same socket for both child and parent.
1-parent) parent calls waitpid to wait for child to exit
2) child successfully echoes
3) child calls exit() which calls some registered atexit handler which shuts down the shared connection
4) parent wakes up, sees client was successful, tries to use the same connection socket which was just shut down.

I don't know how this test worked ever. But most likely there was a race between the SIGCHLD and the exit handlers, and that may have been fixed in recent libc's. I have a small patch to use _exit() in the child forked test part that fixes the issue on my Ubuntu 16.04 box.
